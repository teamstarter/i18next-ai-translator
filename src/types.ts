export interface Config {
    apiKey: string;
    referenceFile?: string;
    bundleReferenceFolder?: string;
    configFile?: string;
  }