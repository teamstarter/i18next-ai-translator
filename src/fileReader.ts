import fs from 'fs';
import path from 'path';
import debug from 'debug';
import { TranslationFile, UntranslatedKey } from './types';

const log = debug('i18next-ai-translator:fileReader');

export function readTranslationFiles(localesFolderPath: string): TranslationFile[] {
  log('Reading translation files from: %s', localesFolderPath);

  if (!fs.existsSync(localesFolderPath)) {
    log('Error: Folder does not exist: %s', localesFolderPath);
    throw new Error(`The folder ${localesFolderPath} does not exist`);
  }

  const files = fs.readdirSync(localesFolderPath);
  log('Found %d files in folder', files.length);

  const jsonFiles = files.filter(file => file.endsWith('.json'));
  log('Filtered to %d JSON files', jsonFiles.length);

  const translationFiles: TranslationFile[] = [];

  for (const file of jsonFiles) {
    const filePath = path.join(localesFolderPath, file);
    const locale = path.basename(file, '.json');

    log('Reading file: %s (locale: %s)', file, locale);

    const content = fs.readFileSync(filePath, 'utf-8');
    const translations = JSON.parse(content);

    translationFiles.push({
      locale,
      filePath,
      translations,
    });

    log('Successfully loaded %s with %d root keys', file, Object.keys(translations).length);
  }

  log('Total translation files loaded: %d', translationFiles.length);
  return translationFiles;
}

function findKeysRecursive(
  obj: Record<string, unknown>,
  currentPath: string[] = [],
  untranslatedValue: string
): string[][] {
  const untranslatedPaths: string[][] = [];

  for (const [key, value] of Object.entries(obj)) {
    const newPath = [...currentPath, key];

    if (typeof value === 'object' && value !== null) {
      const nestedPaths = findKeysRecursive(
        value as Record<string, unknown>,
        newPath,
        untranslatedValue
      );
      untranslatedPaths.push(...nestedPaths);
    } else if (value === untranslatedValue) {
      log('Found untranslated key: %s', newPath.join('.'));
      untranslatedPaths.push(newPath);
    }
  }

  return untranslatedPaths;
}

export function findUntranslatedKeys(
  translationFiles: TranslationFile[],
  untranslatedValue: string = '__STRING_NOT_TRANSLATED__'
): UntranslatedKey[] {
  log('Searching for untranslated keys with value: "%s"', untranslatedValue);
  log('Processing %d translation files', translationFiles.length);

  const untranslatedKeys: UntranslatedKey[] = [];

  for (const file of translationFiles) {
    log('Analyzing file: %s (locale: %s)', file.filePath, file.locale);

    const keyPaths = findKeysRecursive(file.translations, [], untranslatedValue);
    log('Found %d untranslated keys in %s', keyPaths.length, file.locale);

    for (const keyPath of keyPaths) {
      untranslatedKeys.push({
        keyPath: keyPath,
        locale: file.locale,
        filePath: file.filePath,
      });
    }
  }

  log('Total untranslated keys found: %d', untranslatedKeys.length);
  return untranslatedKeys;
}
