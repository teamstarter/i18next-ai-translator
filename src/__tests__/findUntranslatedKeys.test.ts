import { findUntranslatedKeys } from '../fileReader';
import { TranslationFile } from '../types';

describe('findUntranslatedKeys', () => {
  // Test 1 : find a single untranslated key
  it('should find a single untranslated key', () => {
    // create a test file
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          welcome: 'Welcome',
          login: '__STRING_NOT_TRANSLATED__',
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toHaveLength(1);
    expect(result[0].keyPath).toEqual(['login']);
    expect(result[0].locale).toBe('en');
  });

  // Test 2 : find multiple untranslated keys
  it('should find multiple untranslated keys', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          welcome: 'Welcome',
          login: '__STRING_NOT_TRANSLATED__',
          logout: '__STRING_NOT_TRANSLATED__',
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toHaveLength(2);
    expect(result[0].keyPath).toEqual(['login']);
    expect(result[1].keyPath).toEqual(['logout']);
  });

  // Test 3 : find untranslated key in nested object
  it('should find untranslated key in nested object', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          welcome: 'Welcome',
          profile: {
            title: '__STRING_NOT_TRANSLATED__',
          },
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toHaveLength(1);
    expect(result[0].keyPath).toEqual(['profile', 'title']);
  });

  // Test 4 : find deeply nested untranslated key
  it('should find untranslated key deeply nested', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          profile: {
            settings: {
              privacy: '__STRING_NOT_TRANSLATED__',
            },
          },
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toHaveLength(1);
    expect(result[0].keyPath).toEqual(['profile', 'settings', 'privacy']);
  });

  // Test 5 : no untranslated keys
  it('should return empty array when all keys are translated', () => {
    const files: TranslationFile[] = [
      {
        locale: 'fr',
        filePath: '/test/fr.json',
        translations: {
          welcome: 'Bienvenue',
          home: 'Accueil',
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toEqual([]);
  });

  // Test 6 : multiple files with untranslated keys
  it('should find untranslated keys across multiple files', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          login: '__STRING_NOT_TRANSLATED__',
        },
      },
      {
        locale: 'es',
        filePath: '/test/es.json',
        translations: {
          logout: '__STRING_NOT_TRANSLATED__',
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toHaveLength(2);
    expect(result[0].locale).toBe('en');
    expect(result[0].keyPath).toEqual(['login']);
    expect(result[1].locale).toBe('es');
    expect(result[1].keyPath).toEqual(['logout']);
  });

  // Test 7 : empty translation object
  it('should handle empty translation object', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {},
      },
    ];

    const result = findUntranslatedKeys(files);

    expect(result).toEqual([]);
  });

  // Test 8 : custom untranslated value
  it('should work with custom untranslated value', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          welcome: 'Welcome',
          login: 'CUSTOM_NOT_TRANSLATED',
        },
      },
    ];

    // pass our custom value as 2nd parameter
    const result = findUntranslatedKeys(files, 'CUSTOM_NOT_TRANSLATED');

    expect(result).toHaveLength(1);
    expect(result[0].keyPath).toEqual(['login']);
  });

  // Test 9 complex nested structure
  it('should match snapshot for complex nested structure', () => {
    const files: TranslationFile[] = [
      {
        locale: 'en',
        filePath: '/test/en.json',
        translations: {
          welcome: 'Welcome',
          login: '__STRING_NOT_TRANSLATED__',
          logout: '__STRING_NOT_TRANSLATED__',
          home: 'Home',
          profile: {
            title: '__STRING_NOT_TRANSLATED__',
            settings: {
              privacy: '__STRING_NOT_TRANSLATED__',
            },
          },
        },
      },
    ];

    const result = findUntranslatedKeys(files);

    //the complete structure of the found keys
    expect(result).toMatchSnapshot();
  });
});
