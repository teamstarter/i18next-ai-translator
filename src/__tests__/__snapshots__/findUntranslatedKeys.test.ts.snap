// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`findUntranslatedKeys should match snapshot for complex nested structure 1`] = `
[
  {
    "filePath": "/test/en.json",
    "keyPath": [
      "login",
    ],
    "locale": "en",
  },
  {
    "filePath": "/test/en.json",
    "keyPath": [
      "logout",
    ],
    "locale": "en",
  },
  {
    "filePath": "/test/en.json",
    "keyPath": [
      "profile",
      "title",
    ],
    "locale": "en",
  },
  {
    "filePath": "/test/en.json",
    "keyPath": [
      "profile",
      "settings",
      "privacy",
    ],
    "locale": "en",
  },
]
`;
